src/Razor/src/Microsoft.NET.Sdk.Razor/DiscoverMvcApplicationParts.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.AspNetCore.Razor.Tasks
+{
+    public class DiscoverMvcApplicationParts : Task
+    {
+        [Required]
+        public ITaskItem[] MvcAssemblyNames { get; set; }
+
+        [Required]
+        public ITaskItem[] ResolvedReferences { get; set; }
+
+        [Output]
+        public string[] ApplicationPartAssemblyNames { get; set; }
+
+        public override bool Execute()
+        {
+            var referenceItems = new List<ResolveReferenceItem>();
+            foreach (var item in ResolvedReferences)
+            {
+                const string FusionNameKey = "FusionName";
+                var fusionName = item.GetMetadata(FusionNameKey);
+                if (string.IsNullOrEmpty(fusionName))
+                {
+                    Log.LogError($"Missing required metadata '{FusionNameKey}' for '{item.ItemSpec}.");
+                    return false;
+                }
+
+                var assemblyName = new AssemblyName(fusionName).Name;
+                referenceItems.Add(new ResolveReferenceItem
+                {
+                    AssemblyName = assemblyName,
+                    IsSystemReference = item.GetMetadata("IsSystemReference") == "true",
+                    Path = item.ItemSpec,
+                });
+            }
+
+            var mvcAssemblyNames = MvcAssemblyNames.Select(s => s.ItemSpec).ToList();
+
+            var provider = new ReferenceResolver(mvcAssemblyNames, referenceItems);
+            var assemblyNames = provider.ResolveAssemblies();
+
+            ApplicationPartAssemblyNames = assemblyNames.Count > 0 ?
+                assemblyNames.ToArray() :
+                Array.Empty<string>();
+
+            return !Log.HasLoggedErrors;
```

Comment:
Will this actually happen? it's probably better if you just return true here right? Or do you have a reason to expect different.

Reply:
I really like and recommend this pattern. _If_ you're careful it's slightly inefficient but harmless. When someone makes a small mistake (logs an error but a bug keeps the error from propagating all the way back out to the `Execute()` return value), it prevents the super confusing "build succeeded with errors" case.

(In hindsight, it should have been `void Execute()` with logging an error the way to fail it. Alas.)

Reply:
Zoinks! I've been overruled then 🤣